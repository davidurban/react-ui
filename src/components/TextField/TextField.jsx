import PropTypes from 'prop-types';
import React from 'react';
import styles from './TextField.scss';

const TextField = (props) => {
  const {
    changeHandler,
    description,
    error,
    fieldId,
    isLabelVisible,
    ...otherProps
  } = props;

  let labelClass = styles.label;
  if (isLabelVisible) {
    if (props.required) {
      labelClass = styles.isLabelRequired;
    }
  } else {
    labelClass = styles.isLabelHidden;
  }

  return (
    <div className={styles.root}>
      <label htmlFor={fieldId}>
        <div className={labelClass}>
          {props.label}
        </div>
        <input
          {...otherProps}
          id={fieldId}
          onChange={changeHandler}
          className={error ? styles.isInputInvalid : styles.input}
        />
      </label>
      {description && (
        <div className={styles.description}>
          {description}
        </div>
      )}
      {error && (
        <div className={styles.error}>
          {error}
        </div>
      )}
    </div>
  );
};

TextField.defaultProps = {
  changeHandler: null,
  description: null,
  disabled: false,
  error: null,
  isLabelVisible: true,
  placeholder: null,
  required: false,
  type: 'text',
  value: undefined,
};

TextField.propTypes = {
  changeHandler: PropTypes.func,
  description: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  fieldId: PropTypes.string.isRequired,
  isLabelVisible: PropTypes.bool,
  label: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  required: PropTypes.bool,
  type: PropTypes.oneOf(['email', 'number', 'password', 'tel', 'text']),
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export default TextField;
