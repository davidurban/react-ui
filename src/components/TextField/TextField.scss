@import '../Theme/mixins';
@import '../Theme/variables';

.root {
  position: relative;
  display: block;
  box-sizing: border-box;
  padding: 0 0.25rem 1.75rem;
}

.label {
  display: inline-block;
  margin-bottom: 0.25rem;
}

.isLabelRequired {
  composes: label;

  &::after {
    content: ' *';
  }
}

.isLabelHidden {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border: 0;
}

.input {
  @include rui-custom-inputs(1rem);
  @include rui-form-fields-appearance(2.5rem, $rui-form-field-border-color);
  @include rui-input-border-bottom-animation($rui-input-border-color);

  display: block;
}

.input[readonly] {
  border: 0;
}

.input:disabled {
  @include rui-disabled-form-fields();
}

.isInputInvalid {
  composes: input;
  border-bottom: $rui-border-width solid $rui-input-invalid-color;
  background: linear-gradient(to bottom, transparent 95%, $rui-input-invalid-color 95%);
}

.description,
.error {
  font-size: 0.8rem;
}

.error {
  position: absolute;
  bottom: 0.5rem;
  left: 0.25rem;
  color: $rui-input-invalid-color;
}

