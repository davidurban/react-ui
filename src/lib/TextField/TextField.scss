@import '../Theme/mixins';
@import '../Theme/variables';
@import '../Theme/root';

.root {
  position: relative;
  display: flex;
  flex-direction: column;
  box-sizing: border-box;
  padding: 0 0.25rem 0 0;
  margin-bottom: 1rem;
}

.label {
  display: inline-block;
  margin-bottom: 0.25rem;
}

.isLabelRequired {
  composes: label;

  &::after {
    content: ' *';
  }
}

.isLabelHidden {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border: 0;
}

.input {
  @include rui-custom-inputs($rui-input-height);
  @include rui-form-fields-appearance(100%, $rui-form-field-border-color);
  @include rui-input-border-bottom-animation($rui-input-border-color);

  display: block;
  font-family: var(--font-family);
}

.input[readonly] {
  border: 0;
}

.input:disabled {
  @include rui-disabled-form-fields();
}

.isInputInvalid {
  composes: input;
  border-bottom: $rui-border-width solid $rui-input-invalid-color;
  background: linear-gradient(to bottom, transparent 95%, $rui-input-invalid-color 95%);
}

.description,
.error {
  font-size: 0.8rem;
}

.error {
  min-height: 1rem;
  color: $rui-input-invalid-color;
}

