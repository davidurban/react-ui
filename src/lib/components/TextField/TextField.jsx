import PropTypes from 'prop-types';
import React from 'react';
import styles from './TextField.scss';

const TextField = (props) => {
  let labelVisibilityClass = '';
  let rootFullWidthClass = '';
  let rootLayoutClass = '';
  let rootRequiredClass = '';
  let rootStateClass = '';
  let rootVariantClass = '';

  if (!props.isLabelVisible) {
    labelVisibilityClass = styles.isLabelHidden;
  }

  if (props.fullWidth) {
    rootFullWidthClass = styles.isRootFullWidth;
  }

  if (props.layout === 'horizontal') {
    rootLayoutClass = styles.rootLayoutHorizontal;
  } else if (props.layout === 'vertical') {
    rootLayoutClass = styles.rootLayoutVertical;
  }

  if (props.required) {
    rootRequiredClass = styles.isRootRequired;
  }

  if (props.state === 'invalid') {
    rootStateClass = styles.isRootStateInvalid;
  } else if (props.state === 'valid') {
    rootStateClass = styles.isRootStateValid;
  } else if (props.state === 'warning') {
    rootStateClass = styles.isRootStateWarning;
  }

  if (props.variant === 'filled') {
    rootVariantClass = styles.rootVariantFilled;
  } else if (props.variant === 'outline') {
    rootVariantClass = styles.rootVariantOutline;
  }

  return (
    <div
      className={(`
        ${styles.root}
        ${rootFullWidthClass}
        ${rootLayoutClass}
        ${rootRequiredClass}
        ${rootStateClass}
        ${rootVariantClass}
      `).trim()}
    >
      <label className={styles.container} htmlFor={props.fieldId}>
        <div
          className={(`
            ${styles.label}
            ${labelVisibilityClass}
          `).trim()}
        >
          {props.label}
        </div>
        <div className={styles.inputContainer}>
          <input
            className={styles.input}
            disabled={props.disabled}
            id={props.fieldId}
            onChange={props.changeHandler}
            placeholder={props.placeholder}
            required={props.required}
            size={props.size}
            type={props.type}
            value={props.value}
          />
          {props.variant === 'filled' && (
            <div className={styles.bottomLine} />
          )}
        </div>
      </label>
      {props.feedback && (
        <div className={styles.feedback}>
          {props.feedback}
        </div>
      )}
      {props.description && (
        <div className={styles.description}>
          {props.description}
        </div>
      )}
    </div>
  );
};

TextField.defaultProps = {
  changeHandler: null,
  description: null,
  disabled: false,
  feedback: null,
  fullWidth: false,
  isLabelVisible: true,
  layout: 'vertical',
  placeholder: null,
  required: false,
  size: null,
  state: null,
  type: 'text',
  value: undefined,
  variant: 'outline',
};

TextField.propTypes = {
  changeHandler: PropTypes.func,
  description: PropTypes.string,
  disabled: PropTypes.bool,
  feedback: PropTypes.string,
  fieldId: PropTypes.string.isRequired,
  fullWidth: PropTypes.bool,
  isLabelVisible: PropTypes.bool,
  label: PropTypes.string.isRequired,
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  placeholder: PropTypes.string,
  required: PropTypes.bool,
  size: PropTypes.number,
  state: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  type: PropTypes.oneOf(['email', 'number', 'password', 'tel', 'text']),
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  variant: PropTypes.oneOf(['filled', 'outline']),
};

export default TextField;
