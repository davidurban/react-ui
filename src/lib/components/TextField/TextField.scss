@import '../../shared/tools/accessibility';
@import '../../shared/tools/forms';

// Foundation
.root {
  @include form-field();
  @include form-field-horizontal-neighbor();
}

.input {
  @include form-field-input();
}

.bottomLine {
  @include form-field-bottom-line();
}

.helperText {
  @include form-field-helper-text();
}

// Variants
.rootVariantFilled .input {
  @include form-field-input-variant(filled);
}

.rootVariantOutline .input {
  @include form-field-input-variant(outline);
}

// States
.isRootRequired .label {
  @include form-field-required-label();
}

.isRootStateInvalid {
  @include form-field-state(invalid);
}

.isRootStateValid {
  @include form-field-state(valid);
}

.isRootStateWarning {
  @include form-field-state(warning);
}

// Invisible label
.isLabelHidden {
  @include accessibility-hide-text();
}

// Layouts
.isRootFullWidth {
  @include form-field-full-width();
  @include form-field-vertical-neighbor();
}

.rootLayoutHorizontal {
  @include form-field-layout-horizontal();
}

.rootLayoutVertical {
  @include form-field-layout-vertical();
}
