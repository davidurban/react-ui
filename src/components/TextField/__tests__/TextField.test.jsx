import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import TextField from '../index';

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(<TextField fieldId="test" />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(<TextField
      label="label"
      disabled
      fieldId="test"
      value="value"
      helpText="some help"
      errors={['some error', 'another error']}
      placeholder="placeholder"
      type="email"
      required
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls onChange()', () => {
    const spy = sinon.spy();
    const component = mount(<TextField
      fieldId="test"
      onChange={spy}
    />);

    component
      .find('input')
      .simulate('change', {
        preventDefault: () => {},
        target: { value: 'test' },
      });
    expect(spy.calledOnce).toEqual(true);
  });
});
