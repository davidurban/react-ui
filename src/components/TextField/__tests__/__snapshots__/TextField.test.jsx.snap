// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<div>
  <TextField
    disabled={false}
    id="test"
    placeholder={null}
    required={false}
    type="text"
  />
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div>
  <label
    htmlFor="test"
  >
    label
  </label>
  <TextField
    disabled={true}
    id="test"
    placeholder="placeholder"
    required={true}
    type="email"
    value="value"
  />
  <div>
    some help
  </div>
  <div
    key="some error"
  >
    some error
  </div>
  <div
    key="another error"
  >
    another error
  </div>
</div>
`;
